File: DesktopCore/src/bean/RadioButtonPanel.java
package bean;

import java.awt.Component;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.Enumeration;

import javax.swing.AbstractButton;
import javax.swing.ButtonGroup;
import javax.swing.JPanel;
import javax.swing.JRadioButton;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;

public class RadioButtonPanel extends JPanel
							  implements ChangeListener
{
	private ButtonGroup bg = new ButtonGroup();
	private JRadioButton selection = null;
	
	@Override
	public Component add(Component arg0)
	{
		if (arg0 instanceof JRadioButton)
		{
			addRadioButton((JRadioButton)arg0);
		}
		
		return super.add(arg0);
	}
	
	@Override
	public void add(Component arg0, Object arg1)
	{
		if (arg0 instanceof JRadioButton)
		{
			addRadioButton((JRadioButton)arg0);
		}
		
		super.add(arg0, arg1);
	}
	
	private void addRadioButton(JRadioButton rdbtn)
	{
		rdbtn.addChangeListener(this);
		
		bg.add(rdbtn);
		if (bg.getButtonCount() == 1)
		{
			//Ensure this is selected
			rdbtn.setSelected(true);
		}
	}
	
	public JRadioButton getSelection()
	{
		return selection;
	}
	public String getSelectionStr()
	{
		return selection.getText();
	}
	public void setSelection(String selectionStr)
	{
		Enumeration<AbstractButton> buttons = bg.getElements();
		while (buttons.hasMoreElements())
		{
			AbstractButton button = buttons.nextElement();
			String buttonText = button.getText();
			if (buttonText.equals(selectionStr))
			{
				button.setSelected(true);
				stateChanged(new ChangeEvent(button));
				break;
			}
		}
	}
	
	public void addActionListener(ActionListener listener)
	{
		Enumeration<AbstractButton> buttons = bg.getElements();
		while (buttons.hasMoreElements())
		{
			AbstractButton button = buttons.nextElement();
			button.addActionListener(listener);
		}
	}
	
	public boolean isEventSource(ActionEvent evt)
	{
		Object source = evt.getSource();
		
		Enumeration<AbstractButton> buttons = bg.getElements();
		while (buttons.hasMoreElements())
		{
			AbstractButton button = buttons.nextElement();
			if (source == button)
			{
				return true;
			}
		}
		
		return false;
	}
	
	@Override
	public void stateChanged(ChangeEvent arg0)
	{
		JRadioButton src = (JRadioButton)arg0.getSource();
		if (src.isSelected())
		{
			selection = src;
		}
	}
}

File: Darts/src/burlton/dartzee/code/core/bean/RadioButtonPanel.kt
package burlton.dartzee.code.core.bean

import java.awt.Component
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import javax.swing.ButtonGroup
import javax.swing.JPanel
import javax.swing.JRadioButton
import javax.swing.event.ChangeEvent
import javax.swing.event.ChangeListener

open class RadioButtonPanel : JPanel(), ChangeListener
{
    private val bg = ButtonGroup()
    var selection: JRadioButton? = null

    override fun add(arg0: Component): Component
    {
        if (arg0 is JRadioButton)
        {
            addRadioButton(arg0)
        }

        return super.add(arg0)
    }

    override fun add(arg0: Component, arg1: Any)
    {
        if (arg0 is JRadioButton)
        {
            addRadioButton(arg0)
        }

        super.add(arg0, arg1)
    }

    private fun addRadioButton(rdbtn: JRadioButton)
    {
        rdbtn.addChangeListener(this)

        bg.add(rdbtn)
        if (bg.buttonCount == 1)
        {
            //Ensure this is selected
            rdbtn.isSelected = true
        }
    }

    fun getSelectionStr() = selection?.text ?: ""

    fun setSelection(selectionStr: String)
    {
        val buttons = bg.elements
        while (buttons.hasMoreElements())
        {
            val button = buttons.nextElement()
            val buttonText = button.text
            if (buttonText == selectionStr)
            {
                button.isSelected = true
                stateChanged(ChangeEvent(button))
                break
            }
        }
    }

    fun addActionListener(listener: ActionListener)
    {
        val buttons = bg.elements
        while (buttons.hasMoreElements())
        {
            val button = buttons.nextElement()
            button.addActionListener(listener)
        }
    }

    fun removeActionListener(listener: ActionListener)
    {
        val buttons = bg.elements
        while (buttons.hasMoreElements())
        {
            val button = buttons.nextElement()
            button.removeActionListener(listener)
        }
    }

    fun isEventSource(evt: ActionEvent?): Boolean
    {
        evt ?: return false

        val source = evt.source

        val buttons = bg.elements
        while (buttons.hasMoreElements())
        {
            val button = buttons.nextElement()
            if (source === button)
            {
                return true
            }
        }

        return false
    }

    override fun stateChanged(arg0: ChangeEvent)
    {
        val src = arg0.source as JRadioButton
        if (src.isSelected)
        {
            selection = src
        }
    }
}
